rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExpCtor.java

```diff
@@ -113,4 +120,155 @@ private static RegExpImpl getImpl() {
+        if (args.length == 0) {
+            throw ScriptRuntime.typeErrorById("msg.not.a.string");
+        }
+
+        Object arg = args[0];
+        // Check if argument is NOT a string - throw TypeError (no coercion per spec)
+        if (arg == null
+                || arg == Undefined.instance
+                || arg == Undefined.SCRIPTABLE_UNDEFINED
+                || !(arg instanceof CharSequence)) {
+            throw ScriptRuntime.typeErrorById("msg.not.a.string");
+        }
+
+        String input = ScriptRuntime.toString(arg);
+        if (input.isEmpty()) {
+            return "";
+        }
+
+        StringBuilder result = new StringBuilder();
+        boolean isFirst = true;
+
+        for (int i = 0; i < input.length(); i++) {
+            char c = input.charAt(i);
+
+            // SPEC STEP 1: Escape initial character if it's a decimal digit or ASCII letter
+            if (isFirst && (isDecimalDigit(c) || isAsciiLetter(c))) {
+                result.append(String.format("\\x%02x", (int) c));
+                isFirst = false;
+                continue;
+            }
+            isFirst = false;
+
+            // SPEC: Surrogates must be escaped as unicode
+            if (Character.isSurrogate(c)) {
+                result.append(String.format("\\u%04x", (int) c));
+                continue;
+            }
+
+            // SPEC STEP 2: Control escapes (Table 64)
+            switch (c) {
+                case '\t':
+                    result.append("\\t");
+                    continue;
+                case '\n':
+                    result.append("\\n");
+                    continue;
+                case '\u000B': // vertical tab
+                    result.append("\\v");
+                    continue;
+                case '\f':
+                    result.append("\\f");
+                    continue;
+                case '\r':
+                    result.append("\\r");
+                    continue;
+            }
+
+            // SPEC STEP 3: Syntax characters (backslash escape)
+            if (isSyntaxCharacter(c)) {
+                result.append('\\').append(c);
+                continue;
+            }
+
+            // SPEC STEP 4-5: Other punctuators, WhiteSpace, LineTerminator (hex/unicode escape)
+            if (isOtherPunctuator(c) || isWhiteSpace(c) || isLineTerminator(c)) {
+                if (c <= 0xFF) {
+                    result.append(String.format("\\x%02x", (int) c));
+                } else {
+                    result.append(String.format("\\u%04x", (int) c));
+                }
+                continue;
+            }
+
+            // SPEC STEP 6: Return code point as-is (including underscore, other letters, digits)
+            result.append(c);
+        }
+
+        return result.toString();
+    }
+
+    /** Check if character is a decimal digit (0-9) */
+    private static boolean isDecimalDigit(char c) {
```

Comment:
I'm sure that we already have implementations of these checks either elsewhere in the codebase, or in the Java Character class. Can you please check first to see if we can reuse those or make them more generic rather than encode this here? I won't be surprised if we need a few new ones but checks like isWhiteSpace and isDecimalDigit are certainly not unique to this method. Thanks!